src/java/picard/util/TestNGUtil.java

```diff
@@ -0,0 +1,88 @@
+package picard.util;
+
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+
+import org.testng.Assert;
+
+import static java.lang.Math.abs;
+
+/**
+ * Small class implementing some utility functions that are useful for test and interfacing with the TestNG framework.
+ */
+public class TestNGUtil {
+    public static abstract class SingleTestUnitTest<TESTCASE_TYPE extends TestNGParameterizable> {
+        @DataProvider(name = "testcases")
+        Object[][] getParams() {
+            return TestNGUtil.generateDataProvider(getTestcases());
+        }
+
+        @Test(dataProvider = "testcases")
+        public void doMetaTest(final TESTCASE_TYPE testcase) {
+            // Delegate to another method to avoid overwriting of annotations.
+            doTest(testcase);
+        }
+
+        abstract public Iterable<TESTCASE_TYPE> getTestcases();
+
+        abstract public void doTest(TESTCASE_TYPE testcase);
+    }
+
+    /**
+     * Interface for exposing an implementation for converting an object into a Object[] appropriate for positional
+     * parameter-passing as performed by TestNG.  Useful for simplifying data sources.
+     */
+    public static class TestNGParameterizable {
+        public Object[] toObjectArray() {
+            return new Object[]{this};
+        }
+    }
+
+    public static Object[][] generateDataProvider(final TestNGParameterizable[] testcases) {
+        return generateDataProvider(Arrays.asList(testcases));
+    }
+
+    public static Object[][] generateDataProvider(final Iterable<? extends TestNGParameterizable> testcases) {
+        final Iterator<? extends TestNGParameterizable> i = testcases.iterator();
+        final List<Object[]> parameterList = new LinkedList<Object[]>();
+        while (i.hasNext()) {
+            parameterList.add(i.next().toObjectArray());
+        }
+        Object[][] parameterArray = new Object[parameterList.size()][];
+        for (int j = 0; j < parameterList.size(); j++) {
+            parameterArray[j] = parameterList.get(j);
+        }
+        return parameterArray;
+    }
+
+    static final double EPSILON = 1e-300; //a constant near the smallest possible positive value representable in double.
```

Comment:
if you want the actual smallest number, use this `Double.longBitsToDouble(0x1L)`. It's `4.9E-324`